source/MaterialXGenOgsXml/GlslFragmentGenerator.cpp

```diff
@@ -0,0 +1,251 @@
+        {
+            variable += OgsXmlGenerator::SAMPLER_SUFFIX;
+        }
+    }
+    return variable;
+}
+
+const string GlslFragmentGenerator::TARGET = "ogsxml";
+
+GlslFragmentGenerator::GlslFragmentGenerator() :
+    GlslShaderGenerator()
+{
+    // Use our custom syntax class
+    _syntax = std::make_shared<GlslFragmentSyntax>();
+
+    // Set identifier names to match OGS.
+    _identifiers[HW::POSITION_WORLD]        = "Pw";
+    _identifiers[HW::POSITION_OBJECT]       = "Pm";
+    _identifiers[HW::NORMAL_WORLD]          = "Nw";
+    _identifiers[HW::NORMAL_OBJECT]         = "Nm";
+    _identifiers[HW::TANGENT_WORLD]         = "Tw";
+    _identifiers[HW::TANGENT_OBJECT]        = "Tm";
+    _identifiers[HW::BITANGENT_WORLD]       = "Bw";
+    _identifiers[HW::BITANGENT_OBJECT]      = "Bm";
+    _identifiers[HW::VERTEX_DATA_INSTANCE]  = "PIX_IN";
+    _identifiers[HW::ENV_IRRADIANCE]        = "u_envIrradianceSampler";
+    _identifiers[HW::ENV_RADIANCE]          = "u_envRadianceSampler";
+}
+
+ShaderGeneratorPtr GlslFragmentGenerator::create()
+{
+    return std::make_shared<GlslFragmentGenerator>();
+}
+
+ShaderPtr GlslFragmentGenerator::generate(const string& name, ElementPtr element, GenContext& context) const
+{
+    resetIdentifiers(context);
+    ShaderPtr shader = createShader(name, element, context);
+
+    ShaderStage& stage = shader->getStage(Stage::PIXEL);
+
+    // Turn on fixed float formatting to make sure float values are
+    // emitted with a decimal point and not as integers, and to avoid
+    // any scientific notation which isn't supported by all OpenGL targets.
+    Value::ScopedFloatFormatting fmt(Value::FloatFormatFixed);
+
+    const ShaderGraph& graph = shader->getGraph();
+
+    // Add global constants and type definitions
+    emitInclude("pbrlib/" + GlslShaderGenerator::LANGUAGE + "/lib/mx_defines.glsl", context, stage);
+    const unsigned int maxLights = std::max(1u, context.getOptions().hwMaxActiveLightSources);
+    emitLine("#define MAX_LIGHT_SOURCES " + std::to_string(maxLights), stage, false);
+    emitLineBreak(stage);
+    emitTypeDefinitions(context, stage);
+
+    // Add all constants
+    const VariableBlock& constants = stage.getConstantBlock();
+    if (!constants.empty())
+    {
+        emitVariableDeclarations(constants, _syntax->getConstantQualifier(), SEMICOLON, context, stage);
+        emitLineBreak(stage);
+    }
+
+    bool lighting = graph.hasClassification(ShaderNode::Classification::SHADER | ShaderNode::Classification::SURFACE) ||
+        graph.hasClassification(ShaderNode::Classification::BSDF);
+
+    // Emit common math functions
+    emitInclude("pbrlib/" + GlslShaderGenerator::LANGUAGE + "/lib/mx_math.glsl", context, stage);
+    emitLineBreak(stage);
+
+    // Emit lighting functions
+    if (lighting)
+    {
+        if (context.getOptions().hwSpecularEnvironmentMethod == SPECULAR_ENVIRONMENT_FIS)
```

Comment:
You can use GlslShaderGenerator::emitSpecularEnvironment() instead here. It includes checking for the none option.

Reply:
That function is not part of my branch, but will add this in later.